ref-arch/real-time-analytics.rst

```diff
@@ -0,0 +1,248 @@
+#####################
+
+Over the last few years we've helped many different kinds of clients use Citus and noticed
+a technical problem many businesses have: running real-time analytics over large streams
+of data.
+
+For example, say you're building an HTTP analytics dashboard. You have enough clients that
+around every millisecond a user hits one of their websites and sends a log record to you.
+You want to ingest all of those records (1000 inserts/sec) and create a dashboard which
+shows your clients things like how many requests to their sites are errors. It's important
+that this data show up with as little latency as possible so your clients can fix problems
+with their sites. It's also useful to show graphs of historical data, however keeping all
+the raw data around forever is prohibitively expensive.
+
+Or maybe you're building an advertising network and want to show clients clickthrough
+rates on their campaigns. Likewise, you want to ingest lots of data with as little latency
+as possible and show both historical and live data on a dashboard.
+
+In this reference architecture we'll demonstrate how to build part of the first example
+but this architecture would work equally well for the second and many other business
+use-cases.
+
+Running It Yourself
+-------------------
+
+There's `a github repo <http://github.com>`_ with scripts and usage instructions. If
+you've gone through our installation instructions for running on either single or multiple
+machines you're ready to try it out. There will be some code snippets in this tutorial
+but the github repo has all the details in one place.
+
+Data Model
+----------
+
+The data we're dealing with is an immutable stream of log data. Here we'll insert directly
+into Citus but it's also common for this data to first be routed through something like
+Kafka. Doing so makes the system a little more resilient to failures, lets the data be
+routed to multiple places (such as a warehouse like redshift), and (once data volumes
+become unmanageable high) makes it a little easier to start pre-aggregating the data
+before inserting.
+
+In this example, the raw data will use the following schema which isn't very realistic as
+far as http analytics go but sufficient for showing off the architecture we have in mind.
+
+.. code-block:: sql
+
+  CREATE TABLE http_requests (
+    zone_id INT,
+    ingest_time TIMESTAMPTZ DEFAULT now(),
+
+    session_id UUID,
+    url TEXT,
+    request_country TEXT,
+    ip_address CIDR,
+
+    status_code INT,
+    response_time_msec INT,
+  )
+  SELECT master_create_distributed_table('http_requests', 'zone_id', 'hash');
+  SELECT master_create_worker_shards('http_requests', 16, 2);
+
+When we call master_create_distributed_table we're telling it to hash-distribute using
+the `zone_id` column. That means dashboard queries will each hit a single shard.
+
+When we call master_create_worker_shards we tell it to create 16 shards, and 2 copies of
+each shard. `We recommend
```

Comment:
It would be more robust to add

```rst
.. _faq_how_to_choose_shard_count:
```

to `faq/faq.rst` before the subheader for that question and then change this link to

```rst
`We recommend <faq_how_to_choose_shard_count>`_
```